testlapack/dlaqp2.go

```diff
@@ -0,0 +1,132 @@
+// Copyright ©2017 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package testlapack
+
+import (
+	"fmt"
+	"math"
+	"math/rand"
+	"sort"
+	"testing"
+
+	"github.com/gonum/blas"
+	"github.com/gonum/blas/blas64"
+)
+
+type Dlaqp2er interface {
+	Dlaqp2(m, n, offset int, a []float64, lda int, jpvt []int, tau, vn1, vn2, work []float64)
+}
+
+func Dlaqp2Test(t *testing.T, impl Dlaqp2er) {
+	rnd := rand.New(rand.NewSource(1))
+	for ti, test := range []struct {
+		m, n, offset int
+	}{
+		{m: 4, n: 3, offset: 0},
+		// {m: 4, n: 3, offset: 2},
+		// {m: 4, n: 3, offset: 4},
+		{m: 10, n: 10, offset: 0},
+		// {m: 10, n: 10, offset: 5},
+		// {m: 10, n: 10, offset: 10},
+	} {
+		m := test.m
+		n := test.n
+		jpiv := make([]int, n)
+
+		const (
+			all = iota
+			some
+			none
+		)
+		for _, free := range []int{all, some, none} {
+			for _, extra := range []int{0, 11} {
+				a := zeros(m, n, n+extra)
+				c := 1
+				for i := 0; i < m; i++ {
+					for j := 0; j < n; j++ {
+						a.Data[i*a.Stride+j] = float64(c)
+						c++
+					}
+				}
+				aCopy := cloneGeneral(a)
+				for j := range jpiv {
+					switch free {
+					case all:
+						jpiv[j] = -1
+					case some:
+						jpiv[j] = rnd.Intn(2) - 1
+					case none:
+						jpiv[j] = j
+					default:
+						panic("bad freedom")
+					}
+				}
+
+				tau := make([]float64, n)
+				vn1 := make([]float64, n)
```

Comment:
Done.

The situation for Dlaqps is more subtle. I can exercise the code for difficult columns, but it results in a panic because the norms are outside the bounds of `vn2`.